Update to Linux 5.11 drivers

dumbbell opened the pull request:
This branch synchronizes `drm-kmod` with Linux 5.11.

See https://github.com/freebsd/drm-kmod/pull/163#issuecomment-1157699164 for instructions to play with this branch.

**Status**:
![](https://progress-bar.dev/100/)

dumbbell commented:
I slightly lied on the progress :-) I cherry-picked only 484 out of 1334 patches (36%). I'm using this branch as of now to see how it behaves on a day-to-day basis.

dumbbell commented:
The backport is completed for a few weeks now. I'm using this on a daily basis with an Intel CannonLake GPU (i915).

And starting from today, it seems to work with an discrete AMD Radeon RX 6700 XT! On top of that, the dGPU is plugged to my laptop using a Razer Core X Chroma Thunderbolt 3 external case :-) I will start to use that dGPU everyday to further test it.

Here is a screenshot using Sway/Wayland and this 6700 XT:

![freebsd-rx-6700-xt](https://user-images.githubusercontent.com/159804/172019952-8b6dd022-62b9-4512-ace3-71c65cbab489.png)

It depends on a FreeBSD src branch I didn't push yet.

evadot commented:
Awesome news.
Any suspend/resume regression ? (either intel or amdgpu works perfectly here on multiple machines)
Also does amdgpu can still be unloaded and re-loaded (works here), i915kms still have some problems sometimes that I haven't been able to find and fix.

nevillehay commented:
Thanks for all your hard work! I have an RX 6500XT. I'd like to test the 5.11 version. 
How can I do this?

dumbbell commented:
I just updated the comment above with the freshly uploaded firmwares update.

nevillehay commented:
Thanks a lot! I'll try this tomorrow or Monday. I take it this won't work on 13 STABLE?

dumbbell commented:
I don't know. You could try to rebase the `freebsd-src`'s `linuxkpi-changes-for-drm-from-5.11` branch on top of `stable/13` and see if it compiles already.

nevillehay commented:
Okay. Thanks.
I have to wait for a new CPU. Should be coming in the next couple of days.
Sorry for the delay. I also have an r9-270x. Should I use my RX 6500xt or
the r9-270x? The 270x works fine with drm-510-kmod.

On Sun, 19 Jun 2022, 00:01 Jean-Sébastien Pédron, ***@***.***>
wrote:

> I don't know. You could try to rebase the freebsd-src's
> linuxkpi-changes-for-drm-from-5.11 branch on top of stable/13 and see if
> it compiles already.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/freebsd/drm-kmod/pull/163#issuecomment-1159479404>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGD624OVVBQQX2KHB3BWFQTVPXQFDANCNFSM5VBHKSUA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

nevillehay commented:
On second thought I'll do everything with the 6500xt installed

On Sun, 19 Jun 2022, 19:00 Neville Goddard, ***@***.***> wrote:

> Okay. Thanks.
> I have to wait for a new CPU. Should be coming in the next couple of days.
> Sorry for the delay. I also have an r9-270x. Should I use my RX 6500xt or
> the r9-270x? The 270x works fine with drm-510-kmod.
>
> On Sun, 19 Jun 2022, 00:01 Jean-Sébastien Pédron, <
> ***@***.***> wrote:
>
>> I don't know. You could try to rebase the freebsd-src's
>> linuxkpi-changes-for-drm-from-5.11 branch on top of stable/13 and see if
>> it compiles already.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/freebsd/drm-kmod/pull/163#issuecomment-1159479404>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AGD624OVVBQQX2KHB3BWFQTVPXQFDANCNFSM5VBHKSUA>
>> .
>> You are receiving this because you commented.Message ID:
>> ***@***.***>
>>
>

dumbbell commented:
Testing the r9-270x would also be a nice baseline test to spot any regressions. But any testing is welcome! So feel free to try whatever is more convenient of course!

nevillehay commented:
Okay. Thanks. Do I need to install anything for compiling or will FreeBSD
14 with xfce desktop with src be enough. Do I need debugging tools?

On Sun, 19 Jun 2022, 20:41 Jean-Sébastien Pédron, ***@***.***>
wrote:

> Testing the r9-270x would also be a nice baseline test to spot any
> regressions. But any testing is welcome! So feel free to try whatever is
> more convenient for you of course!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/freebsd/drm-kmod/pull/163#issuecomment-1159699823>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGD624PBKDITZLFOI2RXQZDVP4BPBANCNFSM5VBHKSUA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

dumbbell commented:
Do you plan to install a new FreeBSD on a new partition or ZFS filesystem? Do you use boot environments?

If you are currently using FreeBSD 13.x, I would recommend to keep this one untouched, in case you hit any issues with 14-CURRENT that would block you. If you do a fresh install or duplicate your 13.x install into a new boot environment, you should be fine either way. It's been a very long time since I used the FreeBSD installer, but you should already have the base compiler available and it will be enough to recompile FreeBSD and the drm+firmwares kernel modules.

If you duplicate your existing install of 13.x, you can use that clone to do the testing with your existing ports and configuration. In my instructions, I only covered the compilation of the kernel, but that's ok: a 14-CURRENT kernel will work with a 13.x userland. As I said earlier, just make sure to not kldload the DRM drivers during boot, just to make sure to not break boot. Load the module manually at first.

nevillehay commented:
Thanks for your quick replies. I'll do a fresh install of FreeBSD 14. I won't have FreeBSD 13 running at all. I have another computer running FreeBSD (actually GhostBSD) 13 stable so if the FreeBSD 14 goes belly up, it's not a problem

dumbbell commented:
As for debugging configuration and tools, you need to enable kernel core dumps. I have `dump_dev="AUTO"` in `/etc/rc.conf`, I think it's enough. You may need to install gdb to get nice stacktraces as well. You can force a kernel crash with `sudo sysctl debug.kdb.panic=1` to test that code dumps are saved correctly. During the next reboot, the kernel core dump will be written to `/var/crash`.

If you get a kernel core dump while testing the driver, I will be interested in the latest `/var/crash/core.txt.$N` file (by default, they are written in round-robin from 0 to 9). That said, depending on when/where the kernel crashes in the video driver, the kernel may just hang (and a long press on the power button is required) or reboot without writing a core dump.

nevillehay commented:
Okay thanks very much for the information. I ordered a Ryzen 7 5800x and a
cooler. They're coming tomorrow. When I'm up and running I'll let you know.


On Sun, 19 Jun 2022, 21:38 Jean-Sébastien Pédron, ***@***.***>
wrote:

> As for debugging configuration and tools, you need to enable kernel core
> dumps. I have dump_dev="AUTO" in /etc/rc.conf, I think it's enough. You
> may need to install gdb to get nice stacktraces as well. You can force a
> kernel crash with sudo sysctl debug.kdb.panic=1 to test that code dumps
> are saved correctly. During the next reboot, the kernel core dump will be
> written to /var/crash.
>
> If you get a kernel core dump while testing the driver, I will be
> interested in the latest /var/crash/core.txt.$N file (by default, they
> are written in round-robin from 0 to 9). That said, depending on when/where
> the kernel crashes in the video driver, the kernel may just hang (and a
> long press on the power button is required) or reboot without writing a
> core dump.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/freebsd/drm-kmod/pull/163#issuecomment-1159714895>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGD624LX4MXUSP6ZJ6WCAHDVP4IELANCNFSM5VBHKSUA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

nevillehay commented:
I got the new CPU, and got everything working fine on FreeBSD 13 stable. I
just installed FreeBSD 14 from June 17 and there seems to be some sort of
regression. Clean install with xfce desktop. I installed drm-kmod and this
pulled in drm-510-kmod I also install gpu -firmware-kmod but during the
boot process when it says loading amdgpu the screen goes black and stays
black. I also put in rc.conf

kld_list="amdgpu"

I'll have to try a different version.

Currently using scfb to get to the desktop

On Mon, 20 Jun 2022, 11:05 Neville Goddard, ***@***.***> wrote:

> Okay thanks very much for the information. I ordered a Ryzen 7 5800x and a
> cooler. They're coming tomorrow. When I'm up and running I'll let you know.
>
>
> On Sun, 19 Jun 2022, 21:38 Jean-Sébastien Pédron, <
> ***@***.***> wrote:
>
>> As for debugging configuration and tools, you need to enable kernel core
>> dumps. I have dump_dev="AUTO" in /etc/rc.conf, I think it's enough. You
>> may need to install gdb to get nice stacktraces as well. You can force a
>> kernel crash with sudo sysctl debug.kdb.panic=1 to test that code dumps
>> are saved correctly. During the next reboot, the kernel core dump will be
>> written to /var/crash.
>>
>> If you get a kernel core dump while testing the driver, I will be
>> interested in the latest /var/crash/core.txt.$N file (by default, they
>> are written in round-robin from 0 to 9). That said, depending on when/where
>> the kernel crashes in the video driver, the kernel may just hang (and a
>> long press on the power button is required) or reboot without writing a
>> core dump.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/freebsd/drm-kmod/pull/163#issuecomment-1159714895>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AGD624LX4MXUSP6ZJ6WCAHDVP4IELANCNFSM5VBHKSUA>
>> .
>> You are receiving this because you commented.Message ID:
>> ***@***.***>
>>
>

nevillehay commented:
No I didn't. I noticed your instructions for that but how do you link
/usr/src to the clone? What is the command?
Steps 1 and 3 went great!
Again thanks for all your work.

On Thu, 23 Jun 2022, 04:04 Jean-Sébastien Pédron, ***@***.***>
wrote:

> Did you link /usr/src to the freebsd-src clone? I think /usr/src contains
> or links to the vanilla FreeBSD sources.
>
> There is a variable to indicate where to find the source tree but I don't
> recall it's name
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/freebsd/drm-kmod/pull/163#issuecomment-1163499305>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGD624JRHM343HZHWAKGYN3VQNPVHANCNFSM5VBHKSUA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

evadot commented:
> Did you link `/usr/src` to the freebsd-src clone? I think `/usr/src` contains or links to the vanilla FreeBSD sources.
> 
> There is a variable to indicate where to find the source tree but I don't recall it's name

export SYSDIR=/path/to/src/sys (the sys is important)

nevillehay commented:
Thanks. Will do it tonight. Do I run this command first, then the others?


On Thu, 23 Jun 2022, 15:29 Emmanuel Vadot, ***@***.***> wrote:

> Did you link /usr/src to the freebsd-src clone? I think /usr/src contains
> or links to the vanilla FreeBSD sources.
>
> There is a variable to indicate where to find the source tree but I don't
> recall it's name
>
> export SYSDIR=/path/to/src/sys (the sys is important)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/freebsd/drm-kmod/pull/163#issuecomment-1164004971>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGD624NJQJOZK7PWC65XY2DVQP75TANCNFSM5VBHKSUA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

dumbbell commented:
Thanks @evadot!

@nevillehay: Yes, you need to run this command before you try to compile drm-kmod. Note that if you change terminal or reboot, you'll have to run it again as the exported variable's scope is limited to a running shell instance.

When I suggesting to link, I had a symlink in mind, so something like `ln -s /path/to/src /usr/src` (assuming `/usr/src` does not exist). But you don't need both, the exported variable will work just fine.

nevillehay commented:
By the way, there's small error here


   1.

   You need to checkout the FreeBSD src branch I mentionned,
   linuxkpi-changes-for-drm-from-5.11
   <https://github.com/dumbbell/freebsd-src/tree/linuxkpi-changes-for-drm-from-5.11>,
   and compile a kernel from that branch:

   git clone -b linuxkpi-changes-for-drm-from-5.11
https://github.com/dumbbell/freebsd-src.git
   cd freebsd-src
   make -j8 buildkernel DEBUG_FLAGS-g

   # This installs the kernel under another name, `kernel.drm`. Thus,
you keep the default kernel
   # in case of trouble.
   sudo make installkernel DEBUG_FLAGS=-g INSTKERNNAME=kernel.drm

   2.

   Should be


You need to checkout the FreeBSD src branch I mentionned,
linuxkpi-changes-for-drm-from-5.11, and compile a kernel from that branch:

git clone -b linuxkpi-changes-for-drm-from-5.11
https://github.com/dumbbell/freebsd-src.git
cd freebsd-src
make -j8 buildkernel DEBUG_FLAGS=-g

# This installs the kernel under another name, `kernel.drm`. Thus, you keep
the default kernel
# in case of trouble.
sudo make installkernel DEBUG_FLAGS=-g INSTKERNNAME=kernel.drm

On Thu, 23 Jun 2022, 19:14 Neville Goddard, ***@***.***> wrote:

> Thanks a million. Will try it tonight. I'm using the RX 6500 XT now on
> scfb. Compiling the kernel went fine. Installing firmware also went fine.
> Hopefully everything will be fine tonight
>
> On Thu, 23 Jun 2022, 17:23 Jean-Sébastien Pédron, <
> ***@***.***> wrote:
>
>> Thanks @evadot <https://github.com/evadot>!
>>
>> @nevillehay <https://github.com/nevillehay>: Yes, you need to run this
>> command before you try to compile drm-kmod. Note that if you change
>> terminal or reboot, you'll have to run it again as the exported variable's
>> scope is limited to a running shell instance.
>>
>> When I suggesting to link, I had a symlink in mind, so something like ln
>> -s /path/to/src /usr/src (assuming /usr/src does not exist). But you
>> don't need both, the exported variable will work just fine.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/freebsd/drm-kmod/pull/163#issuecomment-1164107142>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AGD624I45LXBXQZAF6B5IK3VQQNJ3ANCNFSM5VBHKSUA>
>> .
>> You are receiving this because you were mentioned.Message ID:
>> ***@***.***>
>>
>

nevillehay commented:
There should be an = after DEBUG_FLAGS

On Thu, 23 Jun 2022, 19:17 Neville Goddard, ***@***.***> wrote:

> By the way, there's small error here
>
>
>    1.
>
>    You need to checkout the FreeBSD src branch I mentionned,
>    linuxkpi-changes-for-drm-from-5.11
>    <https://github.com/dumbbell/freebsd-src/tree/linuxkpi-changes-for-drm-from-5.11>,
>    and compile a kernel from that branch:
>
>    git clone -b linuxkpi-changes-for-drm-from-5.11 https://github.com/dumbbell/freebsd-src.git
>    cd freebsd-src
>    make -j8 buildkernel DEBUG_FLAGS-g
>
>    # This installs the kernel under another name, `kernel.drm`. Thus, you keep the default kernel
>    # in case of trouble.
>    sudo make installkernel DEBUG_FLAGS=-g INSTKERNNAME=kernel.drm
>
>    2.
>
>    Should be
>
>
> You need to checkout the FreeBSD src branch I mentionned,
> linuxkpi-changes-for-drm-from-5.11, and compile a kernel from that branch:
>
> git clone -b linuxkpi-changes-for-drm-from-5.11
> https://github.com/dumbbell/freebsd-src.git
> cd freebsd-src
> make -j8 buildkernel DEBUG_FLAGS=-g
>
> # This installs the kernel under another name, `kernel.drm`. Thus, you
> keep the default kernel
> # in case of trouble.
> sudo make installkernel DEBUG_FLAGS=-g INSTKERNNAME=kernel.drm
>
> On Thu, 23 Jun 2022, 19:14 Neville Goddard, ***@***.***> wrote:
>
>> Thanks a million. Will try it tonight. I'm using the RX 6500 XT now on
>> scfb. Compiling the kernel went fine. Installing firmware also went fine.
>> Hopefully everything will be fine tonight
>>
>> On Thu, 23 Jun 2022, 17:23 Jean-Sébastien Pédron, <
>> ***@***.***> wrote:
>>
>>> Thanks @evadot <https://github.com/evadot>!
>>>
>>> @nevillehay <https://github.com/nevillehay>: Yes, you need to run this
>>> command before you try to compile drm-kmod. Note that if you change
>>> terminal or reboot, you'll have to run it again as the exported variable's
>>> scope is limited to a running shell instance.
>>>
>>> When I suggesting to link, I had a symlink in mind, so something like ln
>>> -s /path/to/src /usr/src (assuming /usr/src does not exist). But you
>>> don't need both, the exported variable will work just fine.
>>>
>>> —
>>> Reply to this email directly, view it on GitHub
>>> <https://github.com/freebsd/drm-kmod/pull/163#issuecomment-1164107142>,
>>> or unsubscribe
>>> <https://github.com/notifications/unsubscribe-auth/AGD624I45LXBXQZAF6B5IK3VQQNJ3ANCNFSM5VBHKSUA>
>>> .
>>> You are receiving this because you were mentioned.Message ID:
>>> ***@***.***>
>>>
>>

dumbbell commented:
Thank you! I just fixed the comment.

nevillehay commented:
I saw one message

Driver bug unable to set devclass

nevillehay commented:
But I noticed in sndstat I have 2 HDMI devices

dumbbell commented:
> I saw one message
> 
> Driver bug unable to set devclass

Were you trying to kldload the driver? Could you please share the output of dmesg(8) after doing that?

dumbbell commented:
Just a general comment: after you try to compile the code, kldload the driver or start X.Org or a Wayland compositor and want to report feedback or ask for help, could you please always provide the output of dmesg(8) (also available in `/var/log/messages`), plus the X.Org log file or the output from the Wayland compositor (depending on what applies to you)?

If you get a kernel panic, I'm interested in the most recent `/var/crash/core.txt.*`. If no a core dump is written to disk, then a photo of the kernel stacktrace on screen would be nice as well.

Thank you :-)

nevillehay commented:
Okay. Will do. By the way there's no drm-xxxx-kmod package needed?

On Fri, 24 Jun 2022, 00:47 Jean-Sébastien Pédron, ***@***.***>
wrote:

> Just a general comment: after you try to compile the code, kldload the
> driver or start X.Org or a Wayland compositor and want to report feedback
> or ask for help, could you please always provide the output of dmesg(8)
> (also available in /var/log/messages), plus the X.Org log file or the
> output from the Wayland compositor (depending on what applies to you)?
>
> If you get a kernel panic, I'm interested in the most recent
> /var/crash/core.txt.*. If no a core dump is written to disk, then a photo
> of the kernel stacktrace on screen would be nice as well.
>
> Thank you :-)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/freebsd/drm-kmod/pull/163#issuecomment-1164581041>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGD624JDUNHS2KGCVEZ7D3DVQSBIXANCNFSM5VBHKSUA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

nevillehay commented:
No worries. You'll have them tonight.

On Fri, 24 Jun 2022, 18:34 Jean-Sébastien Pédron, ***@***.***>
wrote:

> I can't help you without the output of dmesg(8) (after kldloading the
> driver) and the X.Org log file unfortunately.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/freebsd/drm-kmod/pull/163#issuecomment-1165391982>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGD624PGCLQQ3VH7QK4GD4TVQV6J7ANCNFSM5VBHKSUA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

nevillehay commented:
Here are some files:
[xorg-amdgpuload.txt](https://github.com/freebsd/drm-kmod/files/8976509/xorg-amdgpuload.txt)
This is the xorg log file with amdgpu in rc.conf
This is an xorg log file without amdgpu in rc.conf
[xorg-noamdgpu.txt](https://github.com/freebsd/drm-kmod/files/8976529/xorg-noamdgpu.txt)

Here is var/log/messages
[messages1.txt](https://github.com/freebsd/drm-kmod/files/8976533/messages1.txt)

dmesg with amdgpu in rc.conf
[dmesg1.txt](https://github.com/freebsd/drm-kmod/files/8976544/dmesg1.txt)

I'll send another dmesg after kldload amdgpu

nevillehay commented:
Here are some files:
[xorg-amdgpuload.txt](https://github.com/freebsd/drm-kmod/files/8976509/xorg-amdgpuload.txt)
This is the xorg log

nevillehay commented:
dmesg file after loading amdgpu
[dmesg-kldload-amdgpu.txt](https://github.com/freebsd/drm-kmod/files/8976669/dmesg-kldload-amdgpu.txt)

nevillehay commented:
I think you're correct
Here is pciconf -lv
[pciconf.txt](https://github.com/freebsd/drm-kmod/files/8977022/pciconf.txt)

And pciconf -lv | grep -B 3 -A 2 display
[pciconf-display.txt](https://github.com/freebsd/drm-kmod/files/8977024/pciconf-display.txt)

dumbbell commented:
It is supported by Linux. The card's PCI ID was added to the driver in the following commit:
https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?id=a2e9b1666ea728aac7107e28d2106fce7e2aba5e

It's part of Linux 5.15.

andreaf74 commented:
> It is supported by Linux. The card's PCI ID was added to the driver in the following commit: https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?id=a2e9b1666ea728aac7107e28d2106fce7e2aba5e
> 
> It's part of Linux 5.15.

Too bad :( , I tried your patch for the same reason as @nevillehay  and I got the same results, so I believe we should wait for linuxkpi to be aligned at least to 5.15.

If needed I'll be more than willing to test future upgrades.

Thanks

dumbbell commented:
Thank you @andreaf74 for taking the time to test!

nevillehay commented:
Thanks for your help everyone!

JonathanxD commented:
First, thanks a lot for your hard work, I've been using FreeBSD in “terminal-only mode” for a couple of months, but couldn't do everything I needed to. This week I found this PR and I was able to clone the fork repos, compile and load the module.

> When I suggesting to link, I had a symlink in mind, so something like ln -s /path/to/src /usr/src (assuming /usr/src does not exist). But you don't need both, the exported variable will work just fine.

It didn't worked for me, I had to make a symlink, otherwise the compiler would complain about declaration conflicts under x86_64 headers, looks like it was importing both the system one and the cloned one, but I'm not sure, because clang was reporting the path as `./x86_64` (relative to CWD) instead of something like `/usr/src/`, and the conflicting one was being reported on the cloned source repo.

I don't know if this is because I'm on FreeBSD 13.1-RELEASE, but shouldn't since I specified the source path, probably there still another env that I should've changed.

Apart from this, everything seems to be working, sometimes I get hard kernel crashes on the first boot, but in the second one it just works, and on “logout” as well. 

Xorg seems to not be working as it should, KDE only starts using `startplasma-x11` when I boot using UEFI, but don't in BIOS mode, also, Xorg doesn't detect radeonsi driver correctly and seems to be falling back to llvmpipe, but I'm not sure if this is real, since without this module I couldn't get any graphical interface working even with llvmpipe, it was only possible with [UEFI GOP](https://wiki.osdev.org/GOP) and [BIOS VESA](https://wiki.osdev.org/VESA). But under Wayland, the experience is just as good as in Linux, Sway loads correctly and `glxinfo` reports my graphics card. Even Firefox shows the right information.

![swappy-20220728_222221](https://user-images.githubusercontent.com/5360060/181683565-29a20dc2-94fd-4186-b807-009803a2b5d0.png)

I'm very excited in seeing this reaching a mature state, and it already looks like it's almost there.

dumbbell commented:
Thank you very much for testing :-)

The only crash I get is when I exit from Sway so far. Perhaps we hit the same issue. Unfortunately, I don't get a kernel core dump. Do you have core dumps for the panics you have?

I spent quite some time trying to configure Intel AMT to have a remote serial console over the network, but I couldn't get it to work on FreeBSD (it works on Windows on the same laptop).

About X.Org vs. Wayland, I still didn't test X.Org yet.

Beside fixing the panics, the last big piece of work is the I2C. The drivers in this branch currently relies on the GPLv2 implementation. I need to add the missing bits to the implementation in FreeBSD base. That's why I can't rebase this branch on top of the latest `master` because the GPLv2 copy was removed.

JonathanxD commented:
> Do you have core dumps for the panics you have?

I do, I can get them anytime, just need to keep restarting until they happen, I don't know if there is a way to get those messages in Plain Text after the fact (I didn't configured a swap partition so I can't use regular kernel debugging techniques, and I don't have that much knowledge on FreeBSD).

I'm attaching the dump as an image, but if you know if there is a way to get those as text, it would be way easier to read.

![crash](https://user-images.githubusercontent.com/5360060/181867279-3b5ff506-9ab3-47fe-b5a0-678435c3b054.jpg)

sescu54 commented:
Any ETA on when this will be ready and pushed to the main repository? :)

dumbbell commented:
> Any ETA on when this will be ready and pushed to the main repository? :)

Hi @sescu54!

No ETA :) I've been quite busy with other things this summer.

What's left to do:
* Fix a panic, perhaps in the i915 driver, which occurs when another panic is being treated. Probably a difference of behavior between the Linux and FreeBSD consoles. So far, I couldn't find the reason.
* Fix a panic in the amdgpu driver triggered when I exit Sway (but it doesn't happen 100% of the time). I didn't look at it yet and it happens rarely now.
* Test with X.Org. Not started yet.

I won't test suspend & resume because it never worked for me, so I can't see regressions :)

Did you test the driver on your hardware?

evadot commented:
> > Any ETA on when this will be ready and pushed to the main repository? :)
> 
> Hi @sescu54!
> 
> No ETA :) I've been quite busy with other things this summer.
> 
> What's left to do:
> 
>     * Fix a panic, perhaps in the i915 driver, which occurs when another panic is being treated. Probably a difference of behavior between the Linux and FreeBSD consoles. So far, I couldn't find the reason.
> 
>     * Fix a panic in the amdgpu driver triggered when I exit Sway (but it doesn't happen 100% of the time). I didn't look at it yet and it happens rarely now.
> 
>     * Test with X.Org. Not started yet.
> 
> 
> I won't test suspend & resume because it never worked for me, so I can't see regressions :)

 Suspend/Resume works on all my hardware on main and 13.1, please try again.

> Did you test the driver on your hardware?

Skyliegirl33 commented:
I have tested this branch on my machine (RX 6600), and while I'm not sure the GPU driver is at fault, one game in particular refuses to launch and the error is simply "DirectX error". Can't find anything in Wine or mesa that seems to be failing. Issue here; https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=264973

dumbbell commented:
@evadot said:
> Suspend/Resume works on all my hardware on main and 13.1, please try again.

I just tested again:
1. From the console, with the `i915kms` DRM driver loaded
2. From the console, with no DRM driver loaded and `debug.acpi.suspend_bounce=1`

In the first case, the screen turns off and a few seconds later, the fans start to spin (but not at their max speed). In the second case, the screen remains on and the fans start too. In both cases, the system doesn't make more progress and isn't responsive anymore.

The laptop is a Dell Precision 5540.

evadot commented:
> @evadot said:
> 
> > Suspend/Resume works on all my hardware on main and 13.1, please try again.
> 
> I just tested again:
> 
>     1. From the console, with the `i915kms` DRM driver loaded
> 
>     2. From the console, with no DRM driver loaded and `debug.acpi.suspend_bounce=1`
> 
> 
> In the first case, the screen turns off and a few seconds later, the fans start to spin (but not at their max speed). In the second case, the screen remains on and the fans start too. In both cases, the system doesn't make more progress and isn't responsive anymore.
> 
> The laptop is a Dell Precision 5540.

Do you have a discrete GPU ?

Skyliegirl33 commented:
> If you didn't try this branch, what driver do you use?

@dumbbell:
Oh, I ported it to work on 13.1. You can find the repos on my GitHub. I did try this one at first, I should mention, so it's not something with mine.

dumbbell commented:
> > @evadot said:
> > > Suspend/Resume works on all my hardware on main and 13.1, please try again.
> > 
> > 
> > I just tested again:
> > ```
> > 1. From the console, with the `i915kms` DRM driver loaded
> > 
> > 2. From the console, with no DRM driver loaded and `debug.acpi.suspend_bounce=1`
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > In the first case, the screen turns off and a few seconds later, the fans start to spin (but not at their max speed). In the second case, the screen remains on and the fans start too. In both cases, the system doesn't make more progress and isn't responsive anymore.
> > The laptop is a Dell Precision 5540.
> 
> Do you have a discrete GPU ?

Yes, an NVIDIA one in the laptop which I never used. The NVIDIA drivers are not installed.

I also have an AMD GPU in an eGPU external case. It was not connected as part of the test.

dumbbell commented:
> **Update:** Now that I think of it, I use `acpi_call` to turn the NVIDIA GPU off so it doesn't consume power from the battery. I will try again without this.

@evadot: After leaving the NVIDIA dGPU alone (powered on but not used), I could suspend/resume the laptop with this branch! I didn't test with DRM 5.10.

I also tested a quick X.Org session and it looks to work (X.Org and KDE start). I need to configure everything to test it further.

evadot commented:
> 

Yeah dGPU was clearly the problem :)
It should work with 5.10 too.

dumbbell commented:
I'm marking this pull request as ready to review.

There are two panics I see from time to time:
* One in the `amdgpu` driver which I can't explicitly reproduce.
* One in the `i915kms` driver which occurs when the kernel panics elsewhere (i915kms manages vt(4) in my case).

The first one may be repaired in a future version of the `amdgpu` driver as Linux 5.12 is the recommended version for AMD Radeon RX 6xxx GPUs.

The second one is probably because of a difference between the Linux console driver and our vt(4) behavior. I will continue to look into it, but I don't consider it a show stopper for wider testing because it only happens as part of another crash.

evadot commented on linuxkpi/bsd/include/linux/mmzone.h:
```diff
@@ -0,0 +1,11 @@
+
+#define MAX_ORDER	11
+
+#endif
```
This should be in base.

dumbbell replied:
Ok

dumbbell replied:
Done.

evadot commented on linuxkpi/gplv2/include/linux/dma-buf-map.h:
```diff
@@ -0,0 +1,266 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
```
No new GPLv2 files please.

dumbbell replied:
Is there a good method to rewrite such a header?

dumbbell replied:
OpenBSD provides a public domain implementation, I will take this one.

dumbbell replied:
Done.

dumbbell commented:
I submitted pull request #207 to fix build again FreeBSD 14-CURRENT 1400071 (there was a change w.r.t. sysctls).

This branch is now rebased on top of that pull request.

evadot commented:
I'll test this branch early next week on the hardware that I have.
Is the branch bisectable ? (i.e. if there was some FreeBSD specific code that break compilation was it commited in the same linux patch etc ... ?)

dumbbell commented:
I will double check tomorrow. What do you prefer? I suppose, having the build fix in the original Linux commit. I can probably split or squash relevant commits.

evadot commented:
> I will double check tomorrow. What do you prefer? I suppose, having the build fix in the original Linux commit. I can probably split or squash relevant commits.

Squashing the fix with the Linux commit is easier and what I did in the past.
I just don't fix if the problem was in Linux but I usually directly apply the fix commit just after (sometimes it's way after).

dumbbell commented:
> Squashing the fix with the Linux commit is easier and what I did in the past.

I squashed the few build fix commits with into the commits they were fixing, and rebased onto the latest `master`.

evadot commented:
You didn't opened reviews for base linuxkpi ?

dumbbell commented:
No, and neither for firmware updates. Should I submit something to Phabricator? Are GitHub PR ok too? I didn't follow if the latter is supported or not.

evadot commented:
> No, and neither for firmware updates. Should I submit something to Phabricator? Are GitHub PR ok too? I didn't follow if the latter is supported or not.

src commit should be done on reviews.freebsd.org for the firmware open a PR on github.com/freebsd/drm-kmod-firmware

dumbbell commented:
Here is the review for the changes to linuxkpi:
https://reviews.freebsd.org/D36869

dumbbell commented:
Here is the firmware update pull request:
freebsd/drm-kmod-firmware#25

nevillehay commented:
Everything compiled ok but it's not working. The module doesn't seem to be
loading properly. My card is Navi 24. There's support for this card? I'll
do a reinstall and try again. I do kldload amdgpu but nothing happens.

On Thu, 23 Jun 2022, 19:50 Jean-Sébastien Pédron, ***@***.***>
wrote:

> Thank you! I just fixed the comment.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/freebsd/drm-kmod/pull/163#issuecomment-1164260908>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGD624NKDPBGPGNMJWN3TXTVQQ6P7ANCNFSM5VBHKSUA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

dumbbell commented:
@nevillehay: Could you pelase share the entry of your GPU in the output of `pciconf -lv`?

Here is an example with my GPU:
```
...
vgapci2@pci0:9:0:0:	class=0x030000 rev=0xc1 hdr=0x00 vendor=0x1002 device=0x73df subvendor=0x1da2 subdevice=0xe445
    vendor     = 'Advanced Micro Devices, Inc. [AMD/ATI]'
    device     = 'Navi 22 [Radeon RX 6700/6700 XT/6750 XT / 6800M]'
    class      = display
    subclass   = VGA
...
```

You can look for your entry by searching for things like "AMD" or "display".

dumbbell commented:
> I'm using the RX 6500 XT now on scfb.

I think this one was added to Linux 5.12.

nevillehay commented:
> @nevillehay: Could you pelase share the entry of your GPU in the output of `pciconf -lv`?
> 
> Here is an example with my GPU:
> 
> ```
> ...
> vgapci2@pci0:9:0:0:	class=0x030000 rev=0xc1 hdr=0x00 vendor=0x1002 device=0x73df subvendor=0x1da2 subdevice=0xe445
>     vendor     = 'Advanced Micro Devices, Inc. [AMD/ATI]'
>     device     = 'Navi 22 [Radeon RX 6700/6700 XT/6750 XT / 6800M]'
>     class      = display
>     subclass   = VGA
> ...
> ```
> 
> You can look for your entry by searching for things like "AMD" or "display".

OK. Can you please wait? I have to put the card in, I'll do it asap. Thanks for your work.

nevillehay commented:
> > @nevillehay: Could you pelase share the entry of your GPU in the output of `pciconf -lv`?
> > Here is an example with my GPU:
> > ```
> > ...
> > vgapci2@pci0:9:0:0:	class=0x030000 rev=0xc1 hdr=0x00 vendor=0x1002 device=0x73df subvendor=0x1da2 subdevice=0xe445
> >     vendor     = 'Advanced Micro Devices, Inc. [AMD/ATI]'
> >     device     = 'Navi 22 [Radeon RX 6700/6700 XT/6750 XT / 6800M]'
> >     class      = display
> >     subclass   = VGA
> > ...
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > You can look for your entry by searching for things like "AMD" or "display".
> 
[pciconf.txt](https://github.com/freebsd/drm-kmod/files/9846340/pciconf.txt)

> OK. Can you please wait? I have to put the card in, I'll do it asap. Thanks for your work.

nevillehay commented:
> > > @nevillehay: Could you pelase share the entry of your GPU in the output of `pciconf -lv`?
> > > Here is an example with my GPU:
> > > ```
> > > ...
> > > vgapci2@pci0:9:0:0:	class=0x030000 rev=0xc1 hdr=0x00 vendor=0x1002 device=0x73df subvendor=0x1da2 subdevice=0xe445
> > >     vendor     = 'Advanced Micro Devices, Inc. [AMD/ATI]'
> > >     device     = 'Navi 22 [Radeon RX 6700/6700 XT/6750 XT / 6800M]'
> > >     class      = display
> > >     subclass   = VGA
> > > ...
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > You can look for your entry by searching for things like "AMD" or "display".
> 
> [pciconf.txt](https://github.com/freebsd/drm-kmod/files/9846340/pciconf.txt)
> 
> > OK. Can you please wait? I have to put the card in, I'll do it asap. Thanks for your work.
[pciconf.txt](https://github.com/freebsd/drm-kmod/files/9846341/pciconf.txt)

nevillehay commented:
The RX 6600 (Navi 23) is supported, correct?

dumbbell commented:
> The RX 6600 (Navi 23) is supported, correct?

Correct. I believe it was added to Linux 5.12. See #208 which is nearly complete.

evadot commented:
Build ok, can you rebase your branch so I can merge this ?
Thanks,

dumbbell commented:
The branch is already rebased on the latest `master`. Did you mean something else?

evadot commented:
> The branch is already rebased on the latest `master`. Did you mean something else?

Looks like it's just github that can't handle big merge request, I'll do this by hand later today or this weekend :)

patmaddox commented:
> 1. You need to checkout the FreeBSD src branch I mentionned, [`linuxkpi-changes-for-drm-from-5.11`](https://github.com/dumbbell/freebsd-src/tree/linuxkpi-changes-for-drm-from-5.11)

@dumbbell that branch no longer exists - has it been merged to freebsd-src?

dumbbell commented:
@patmaddox: Yes, all required changes to linuxkpi for Linux 5.11 were committed to `main`. There are still linuxkpi patches I need to work on for Linux 5.12. See #208.
